packages/codemirror-blocks/src/components/NodeEditable.tsx

```diff
@@ -67,20 +67,21 @@ const NodeEditable = (props: Props) => {
     return { isErrored, initialValue };
   });
 
+  // select and focus the element on mount
+  useEffect(() => {
```

Comment:
Wow, thanks for the exhaustive explanation! In that case, it sounds like we have the following design principles:

1. timeouts of any kind should be used as close to the leaves of the component tree as possible
2. timeouts should probably not be used for anything that isn't strictly DOM dependent 

Would you agree?

Reply:
I would probably say something more general like:

0. Avoid timeouts (and any asynchronous code) if at all possible.
1. Only use timeouts when blocking execution is not acceptable:
    - The computation being done takes more than 30ms, so blocking would cause the page to "freeze up"
    - You actually need to wait before performing the execution (i.e. exponential back-off retries, intentional application behavior) and don't want to freeze the page while waiting
2. Timeout callbacks must not depend on mutable state or must be able to do the right thing when the mutable state has changed. i.e. just because that dom node was there when setTimeout was called, doesn't mean it will be there when the callback executes, so fail gracefully.
3. Code using timeouts should be well documented and provide hooks (ideally using promises) for other code to wait for a timeout to complete.

But asynchronous code is just hard, particularly with shared state, even in languages that provide special constructs for dealing with it. I haven't ever encountered design principles for async code that don't come with a huge number of exceptional cases that end up making the design principles not very useful in practice.